### PR TITLE
fix: query node issues and add test cases

### PR DESCRIPTION
**Pull request by ReebaPaysys**
# SPDX-License-Identifier: Apache-2.0
This pull request introduces a robust SQL query validation and tenant isolation system for SELECT statements, replacing the previous regex-based approach with AST parsing for improved security and flexibility. It adds comprehensive unit tests, leverages the `pgsql-ast-parser` library, and ensures that all user queries are strictly validated, tenant-scoped, and safely parameterized. The changes also provide utility functions for extracting table references, handling template placeholders, and validating sort columns.

**Query validation and AST parsing:**
- Added `pgsql-ast-parser` as a dependency for parsing and analyzing SQL queries at the AST level. (`package.json`, [package.jsonR42](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R42))
- Implemented `validateSelectQuery` in `src/utils/validateQuery.ts` to strictly allow only single SELECT-family queries, reject multiple statements, and handle template placeholders safely. (`src/utils/validateQuery.ts`, [src/utils/validateQuery.tsL1-R42](diffhunk://#diff-c6a237e021b137de0e75cda96886285f776c34ff379b563f9db81a0103e5dad9L1-R42))
- Replaced regex-based validation and table extraction in `executeSelectQuery` with AST-based logic, ensuring queries are parsed, validated, and all referenced tables are extracted for further checks. (`src/services/node.logic.service.ts`, [src/services/node.logic.service.tsL92-R363](diffhunk://#diff-9b8f58fc26d28e5e59f324698ae60da4740d4fcf8f4604b28ab7bdf293917ebcL92-R363))

**Tenant isolation and query modification:**
- Introduced AST-driven tenant condition injection, ensuring tenant filters are applied only where missing, and always in a positionally safe manner (even with subqueries/CTEs). (`src/services/node.logic.service.ts`, [src/services/node.logic.service.tsR11-R204](diffhunk://#diff-9b8f58fc26d28e5e59f324698ae60da4740d4fcf8f4604b28ab7bdf293917ebcR11-R204))
- Changed `resolveTenantColumn` to use parameterized queries for column lookup, improving security and correctness. (`src/services/node.logic.service.ts`, [src/services/node.logic.service.tsL92-R363](diffhunk://#diff-9b8f58fc26d28e5e59f324698ae60da4740d4fcf8f4604b28ab7bdf293917ebcL92-R363))

**Security and validation improvements:**
- All table names extracted from queries are validated using `validateTableName` before any database access, blocking unsafe or forbidden tables. (`src/services/node.logic.service.ts`, [src/services/node.logic.service.tsL92-R363](diffhunk://#diff-9b8f58fc26d28e5e59f324698ae60da4740d4fcf8f4604b28ab7bdf293917ebcL92-R363))
- The system now enforces a hard row limit (`LIMIT 10`) by wrapping the user query, preventing resource exhaustion attacks. (`src/services/node.logic.service.ts`, [src/services/node.logic.service.tsL92-R363](diffhunk://#diff-9b8f58fc26d28e5e59f324698ae60da4740d4fcf8f4604b28ab7bdf293917ebcL92-R363))

**Testing and coverage:**
- Added a comprehensive test suite for `validateSelectQuery`, covering valid/invalid queries, template placeholders, SQL injection attempts, and edge cases. (`__tests__/unit/validateQuery.test.ts`, [__tests__/unit/validateQuery.test.tsR1-R159](diffhunk://#diff-f44c9be93ff409d4ad68efced31a85ceb2e34d7de5dfd2aaf9471a8f39ccec38R1-R159))

**Codebase cleanup and refactoring:**
- Removed the old `validateSystemFunctions` implementation and refactored related logic to use the new AST-based validation. (`src/utils/validateQuery.ts`, [src/utils/validateQuery.tsL1-R42](diffhunk://#diff-c6a237e021b137de0e75cda96886285f776c34ff379b563f9db81a0103e5dad9L1-R42))
- Cleaned up the signature of `executeQueryNodeInDb` to remove the unused `tenantId` parameter. (`src/repositories/configuration/node.repository.ts`, [src/repositories/configuration/node.repository.tsL83](diffhunk://#diff-3f758a776ea420c99854d963cd8fd3e9e0221b25f4db70effd9ee7a7f777cc76L83))

## Why are we doing this?
These changes collectively improve the security, correctness, and maintainability of the query execution path for the `executeQueryNode` API.

## How was it tested?
- [ ] Locally
- [ ] Development Environment
- [ ] Not needed, changes very basic
- [ ] Husky successfully run
- [ ] Unit tests passing and Documentation done

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests covering query validation and node-related operations, including many edge cases, error paths, and tenant-scoped behaviors.

* **New Features**
  * Enhanced SELECT query validation and parsing with clearer error responses.
  * Safer tenant-aware query execution with AST-driven injection, preserved user limits/sorting, and enforced outer result limits.

* **Chores**
  * Added SQL AST parsing dependency.

### CODE REVIEW

File: src/services/node.logic.service.ts [hunk 1]
```diff
@@ -89,68 +187,89 @@ export const executeSelectQuery = async (
...
-    modifiedQuery = modifiedQuery.replace(/;?$/, ' LIMIT 5;');
-  }
+  // 7. Hard-cap rows — strip any user-supplied LIMIT/OFFSET/FETCH and enforce our own cap
+  //    Remove LIMIT with optional OFFSET
+  modifiedQuery = modifiedQuery.replace(/\bLIMIT\s+\d+(\s+OFFSET\s+\d+)?\s*$/i, '').trimEnd();
+  //    Remove standalone OFFSET clause
+  modifiedQuery = modifiedQuery.replace(/\bOFFSET\s+\d+\s*$/i, '').trimEnd();
+  //    Remove FETCH FIRST/NEXT variants (PostgreSQL standard syntax)
+  modifiedQuery = modifiedQuery.replace(/\bFETCH\s+(FIRST|NEXT)\s+\d+\s+ROWS?\s+ONLY\s*$/i, '').trimEnd();
+  modifiedQuery = `${modifiedQuery} LIMIT 10`;
```
[the remark this answers is not in the document]
Comment (ReebaPaysys) [reply src/services/node.logic.service.ts]: @copilot @coderabbitai, how to fix it? Any suggestions?

File: __tests__/unit/node.logic.service.test.ts [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+
+// Mock the repository
+jest.mock('../../src/repositories/configuration/node.repository', () => ({
+  getNodeByIdFromDb: jest.fn(),
+  insertNodesIntoDb: jest.fn(),
+  deleteNodeByIdFromDB: jest.fn(),
+  getAllNodes: jest.fn(),
+  getNodeByName: jest.fn(),
+  executeQueryNodeInDb: jest.fn(),
+}));
```
[the remark this answers is not in the document]
Comment (ReebaPaysys) [reply __tests__/unit/node.logic.service.test.ts:22]: Description updated.

**Comment by ReebaPaysys** [comment]
image

**Review by Justus-at-Tazama — DISMISSED** [review]

**Review by Sandy-at-Tazama — APPROVED** [review]
